extractor/filesystem/os/kernel/module/module.go

```diff
@@ -0,0 +1,282 @@
+	}
+}
+
+// Config returns the configuration of the extractor.
+func (e Extractor) Config() Config {
+	return Config{
+		Stats:            e.stats,
+		MaxFileSizeBytes: e.maxFileSizeBytes,
+	}
+}
+
+// Name of the extractor.
+func (e Extractor) Name() string { return Name }
+
+// Version of the extractor.
+func (e Extractor) Version() int { return 0 }
+
+// Requirements of the extractor.
+func (e Extractor) Requirements() *plugin.Capabilities { return &plugin.Capabilities{} }
+
+// FileRequired returns true if the specified file matches the *.ko file patterns.
+func (e Extractor) FileRequired(api filesystem.FileAPI) bool {
+	path := api.Path()
+
+	if !strings.HasSuffix(filepath.Base(path), ".ko") {
+		return false
+	}
+
+	fileinfo, err := api.Stat()
+	if err != nil {
+		return false
+	}
+
+	if e.maxFileSizeBytes > 0 && fileinfo.Size() > e.maxFileSizeBytes {
+		e.reportFileRequired(path, fileinfo.Size(), stats.FileRequiredResultSizeLimitExceeded)
+		return false
+	}
+
+	e.reportFileRequired(path, fileinfo.Size(), stats.FileRequiredResultOK)
+	return true
+}
+
+func (e Extractor) reportFileRequired(path string, fileSizeBytes int64, result stats.FileRequiredResult) {
+	if e.stats == nil {
+		return
+	}
+	e.stats.AfterFileRequired(e.Name(), &stats.FileRequiredStats{
+		Path:          path,
+		Result:        result,
+		FileSizeBytes: fileSizeBytes,
+	})
+}
+
+// Extract extracts packages from .ko files passed through the scan input.
+func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
+	inventory, err := e.extractFromInput(ctx, input)
+
+	if e.stats != nil {
+		var fileSizeBytes int64
+		if input.Info != nil {
+			fileSizeBytes = input.Info.Size()
+		}
+		e.stats.AfterFileExtracted(e.Name(), &stats.FileExtractedStats{
+			Path:          input.Path,
+			Result:        filesystem.ExtractorErrorToFileExtractedResult(err),
+			FileSizeBytes: fileSizeBytes,
+		})
+	}
+	return inventory, err
+}
+
+func (e Extractor) extractFromInput(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
+	pkgs := []*extractor.Inventory{}
+
+	m, err := osrelease.GetOSRelease(input.FS)
+	if err != nil {
+		log.Errorf("osrelease.ParseOsRelease(): %v", err)
+	}
+
+	// Return if canceled or exceeding deadline.
+	if err := ctx.Err(); err != nil {
+		return pkgs, fmt.Errorf("%s halted at %q because of context error: %v", e.Name(), input.Path, err)
+	}
+
+	data, err := io.ReadAll(input.Reader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read input: %v", err)
+	}
+
+	elfFile, err := elf.NewFile(bytes.NewReader(data))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse ELF file: %v", err)
+	}
+
+	section := elfFile.Section(".modinfo")
+	if section == nil {
+		return nil, fmt.Errorf("no .modinfo section found")
+	}
+
+	sectionData, err := section.Data()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read .modinfo section: %v", err)
+	}
+
+	var (
+		moduleName    string
```

Comment:
Why not just init an empty Metadata struct here and set the fields of that later on?